lib/upnp/upnp.go

```diff
@@ -204,7 +232,7 @@ loop:
 			break
 		}
 
-		igds, err := parseResponse(ctx, deviceType, resp[:n])
+		igds, err := parseResponse(ctx, deviceType, udpAddr, resp[:n], intf)
```

Comment:
This is the origin of the error:

```
upnp.go:242: INFO: UPnP parse: XML syntax error on line 96: unexpected EOF
```

Is the `resp` in line 213 problematic?

Reply:
I see. I don't think it's line 213 (the buffer is 64 KiB in size, that should be enough for a response like that), but still, this means that the error is likely to be in the SSDP part. SSDP runs over UDP so weird behavior is definitely possible (UDP doesn't give many guarantees). In particular, I think the response could be fragmented before being sent and sometimes only the first IP packet could arrive in time. Maybe increasing the timeout set above helps...

There's probably no reasonable "proper" solution to this, since UDP is connectionless so we can't really know if the message is over or if there's a network error that caused the end of the message to disappear.

Reply:
Are you sure? The igddesc.xml seems cut off and I thought that one is requested via HTTP?

Reply:
Oh, you're right. I think I just went immediately to the part of the code with UDP, since that would've been one obvious way this might happen. Too be honest, I have no real explanation right now for this to be happening via a normal HTTP connection though.

Reply:
The response is also cut in Wireshark. I think we can assume that my device is buggy.

Reply:
I implemented retrying the request now, I think using time.Sleep is fine here because this is run as a goroutine per  device anyway so shouldn't really hold anything up.

Reply:
Are we running IPv4 and IPv6 in parallel? That would explain why the IPv6 of the router ends up in the IPv4 igddesc response.

Reply:
Yes, we were already running discovery for IGDv1 and IGDv2 in parallel beforehand so I made it run all 4 requests (IPv6 and IPv4 + IGDv1 and IGDv2) in parallel.  I've split those up now so it waits for IPv4 discovery before starting IPv6 (and it also doesn't try to discover IPv6 IGDv2 devices in the first place now, while before it only discarded them later on)

Reply:
Maybe it worked previously because Syncthing only requested via IPv4. This all smells like my router isn't fully threadsafe regarding its representation URL. Responding with an IPv6 on an IPv4 response is a red flag. I fear the requesting IP is stored in an unsafe way and we trigger a bug now.

Reply:
That's probably it, but no matter what it is, there's probably nothing more that we can do apart from working around it with retries and doing the IPv6 request after the IPv4 one is done.

Reply:
Exactly.

Reply:
Lol. I think I figured it out. The router uses the IPv4 to calculate the size of the response. By falsely inserting the IPv6 into the representationURL this calculation is  off by 12 characters which causes the XML document to be truncated. This would also explains why it's always truncated at the same position.

@maxi0604 nothing for you to fix. I will contact the vendor of my device.